### PR TITLE
feat: create a mongodb cluster

### PR DESCRIPTION
**Pull request by aneojgurhem**
Improve the current MongoDB deployment to support clustering.

### CODE REVIEW

File: infrastructure/modules/onpremise-storage/mongodb/configmap.tf [hunk 1]
```diff
@@ -15,6 +15,61 @@ db.createUser(
...
+
+# TODO: put this file at another place
+CLUSTER_KEY=/data/db/cluster.key
+
+if [ ! -e $CLUSTER_KEY ] ; then
+  cp /cluster/cluster.key $CLUSTER_KEY
+  chmod 400 $CLUSTER_KEY
+fi
+
+/usr/local/bin/docker-entrypoint.sh mongod \
```
Comment (lemaitre-aneo) [inline infrastructure/modules/onpremise-storage/mongodb/configmap.tf:42]: You might need need to put it as the last command with an exec, in order to ensure signals are properly propagated. The rs init "job" would be started in background with a sleep.

File: kubectl
Comment (aneojgurhem) [inline kubectl:1]: that should not be there. Consider rebasing to remove it from history

File: infrastructure/modules/onpremise-storage/mongodb/locals.tf [hunk 2]
```diff
@@ -1,5 +1,9 @@
...
-  # ActiveMQ node selector
-  node_selector_keys   = keys(var.mongodb.node_selector)
-  node_selector_values = values(var.mongodb.node_selector)
+  node_selector_keys         = keys(var.mongodb.node_selector)
+  node_selector_values       = values(var.mongodb.node_selector)
+  replicas                   = toset([for s in range(var.mongodb.replicas_number) : tostring(s)])
+  mongodb_port               = kubernetes_service.mongodb.0.spec.0.type == "NodePort" ? kubernetes_service.mongodb.0.spec.0.port.0.node_port : kubernetes_service.mongodb.0.spec.0.port.0.port
+  mongodb_dns                = "${kubernetes_service.mongodb.0.metadata.0.name}.${kubernetes_service.mongodb.0.metadata.0.namespace}"
+  mongodb_url                = "mongodb://${local.mongodb_dns}:${local.mongodb_port}"
+  mongodb_number_of_replicas = var.mongodb.replicas_number
```
Comment (lzianekhodja) [inline infrastructure/modules/onpremise-storage/mongodb/locals.tf:8]: You can use directly "var.mongodb.replicas_number"

File: infrastructure/modules/onpremise-storage/mongodb/mongodb.tf [hunk 3]
```diff
@@ -126,27 +142,29 @@ resource "kubernetes_deployment" "mongodb" {
 
 # Kubernetes MongoDB service
 resource "kubernetes_service" "mongodb" {
+  for_each = local.replicas
   metadata {
-    name      = kubernetes_deployment.mongodb.metadata.0.name
-    namespace = kubernetes_deployment.mongodb.metadata.0.namespace
+    name      = "mongodb-${each.key}"
```
Comment (lzianekhodja) [inline infrastructure/modules/onpremise-storage/mongodb/mongodb.tf:147]:
```suggestion
    name      = kubernetes_deployment.mongodb[each.key].metadata.0.name
```

File: infrastructure/modules/onpremise-storage/mongodb/mongodb.tf [hunk 4]
```diff
@@ -126,27 +142,29 @@ resource "kubernetes_deployment" "mongodb" {
...
+    namespace = var.namespace
```
Comment (lzianekhodja) [inline infrastructure/modules/onpremise-storage/mongodb/mongodb.tf:148]:
```suggestion
    namespace = kubernetes_deployment.mongodb[each.key].metadata.0.namespace
```

File: infrastructure/quick-deploy/localhost/all/storage.tf [hunk 5]
```diff
@@ -137,12 +138,12 @@ locals {
       ssl_host     = "127.0.0.1"
     } : null
     mongodb = {
-      url                = module.mongodb.url
```
Comment (lzianekhodja) [inline infrastructure/quick-deploy/localhost/all/storage.tf:140]: Why did you delete it ?

File: infrastructure/modules/onpremise-storage/mongodb/locals.tf [hunk 6]
```diff
@@ -1,5 +1,8 @@
 locals {
-  # ActiveMQ node selector
   node_selector_keys   = keys(var.mongodb.node_selector)
   node_selector_values = values(var.mongodb.node_selector)
+  replicas             = toset([for s in range(var.mongodb.replicas_number) : tostring(s)])
```
Comment (lzianekhodja) [inline infrastructure/modules/onpremise-storage/mongodb/locals.tf:4]: Delete

**Review by lzianekhodja — APPROVED** [review]

**Review by lzianekhodja — DISMISSED** [review]

File: infrastructure/modules/armonik/partitions-in-database.tf
Comment (lemaitre-aneo) [inline infrastructure/modules/armonik/partitions-in-database.tf:1]: I thought we didn't need this file?!

File: infrastructure/modules/onpremise-storage/mongodb/configmap.tf (hunk 1 above)
Comment (lemaitre-aneo) [reply infrastructure/modules/onpremise-storage/mongodb/configmap.tf:42]: ping?

File: infrastructure/modules/onpremise-storage/mongodb/mongodb.tf [hunk 7]
```diff
@@ -1,12 +1,13 @@
 # Kubernetes MongoDB deployment
 resource "kubernetes_deployment" "mongodb" {
+  for_each = local.replicas
```
Comment (lemaitre-aneo) [inline infrastructure/modules/onpremise-storage/mongodb/mongodb.tf:3]: Wouldn't it be better to have a count?

File: infrastructure/quick-deploy/aws/storage/outputs.tf [hunk 8]
```diff
@@ -22,7 +22,8 @@ output "storage_endpoint_url" {
       kms_key_id            = module.s3_os[0].kms_key_id
     } : null
     mongodb = {
-      url = module.mongodb.url
+      url                = module.mongodb.url
+      number_of_replicas = var.mongodb.replicas_number
```
Comment (lemaitre-aneo) [inline infrastructure/quick-deploy/aws/storage/outputs.tf:26]: You need some consistency in the naming.

**Comment by lemaitre-aneo** [comment]
You should add a line in the CHANGELOG

**Review by lzianekhodja — APPROVED** [review]

File: infrastructure/quick-deploy/aws/storage/outputs.tf (hunk 8 above)
Comment (melflitty) [reply infrastructure/quick-deploy/aws/storage/outputs.tf:26]: "replicas_number" and "number_of-replicas" refer to the same value. But, I have used "replicas_number" where it is used as input variable and "number_of_replicas" where it is used as output variable. Having two names is not required, I did it to avoid the confusion.

File: infrastructure/modules/armonik/partitions-in-database.tf
Comment (melflitty) [reply infrastructure/modules/armonik/partitions-in-database.tf:1]: We(Me and Imad) encountered problems with deployment. Lilia and Jérôme suggested to re-add this file to have the job "partitions in database"
